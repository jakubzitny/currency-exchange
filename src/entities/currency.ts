// export type Currency = 'USD' | 'GBP' | 'EUR'

export enum Currency {
  USD = 'USD',
  GBP = 'GBP',
  EUR = 'EUR',
}
