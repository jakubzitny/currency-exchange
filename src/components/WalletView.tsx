import React from 'react'
import { List } from 'immutable'

import Wallet from '../entities/wallet'
import { Currency } from '../entities/currency'
import { validateAmount } from '../utils/validation-utils'

import ConversionPart from './ConversionPart'
import Switcher from './Switcher'

import styles from '../styles/WalletView.module.css'

type Props = {
  wallet: Wallet
  wallets: List<Wallet>
  exchangeAmount: number
  exchangeRate: number
  exchangeCurrency: Currency
  type: 'source' | 'target'
  onAmountChange: (amount: number) => void
  onWalletChange: (currency: Currency) => void
}

const WalletView: React.FC<Props> = (props: Props) => {
  const {
    exchangeAmount,
    exchangeCurrency,
    exchangeRate,
    type,
    wallet,
    wallets,
  } = props

  const handleAmountChangeRequest = (
    e: React.ChangeEvent<HTMLInputElement>
  ) => {
    e.persist()
    const validNextAmount = validateAmount(e.target.value)
    if (validNextAmount === false) {
      return
    }

    props.onAmountChange(validNextAmount)
  }

  return (
    <div className={styles.container}>
      <ConversionPart
        currency={wallet.currency}
        walletAmount={wallet.amount}
        exchangeAmount={exchangeAmount}
        rate={exchangeRate}
        rateCurrency={exchangeCurrency}
        type={type}
        onAmountChangeRequest={handleAmountChangeRequest}
      />
      <Switcher
        type={type}
        wallets={wallets}
        activeWallet={wallet.currency}
        onSwitchRequest={props.onWalletChange}
      />
    </div>
  )
}

export default WalletView
