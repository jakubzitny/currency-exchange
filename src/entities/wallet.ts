import { Record } from 'immutable'
import { Currency } from './currency'

interface Defaults {
  currency: Currency
  amount: number
}

const defaults: Defaults = {
  currency: Currency.EUR,
  amount: 0,
}

export default class Wallet extends Record<Defaults>(defaults) {}
