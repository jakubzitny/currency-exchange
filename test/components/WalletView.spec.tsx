import chai, { expect } from 'chai'
import chaiSinon from 'sinon-chai'
import chaiJestSnapshot from 'chai-jest-snapshot'
import { shallow } from 'enzyme'
import React from 'react'
import { spy } from 'sinon'
import { List } from 'immutable'

import { Currency } from '../../src/entities/currency'
import Wallet from '../../src/entities/wallet'
import WalletView from '../../src/components/WalletView'

chai.use(chaiJestSnapshot)
chai.use(chaiSinon)

describe('WalletView', () => {
  const createWalletView = (
    props: Object = {},
    type: 'source' | 'target' = 'source'
  ) => {
    const wallet = new Wallet({ currency: Currency.EUR, amount: 0 })
    const wallet2 = new Wallet({ currency: Currency.USD, amount: 0 })
    const defaultProps = {
      wallet,
      wallets: List.of(wallet, wallet2),
      exchangeAmount: 0,
      exchangeRate: 1,
      exchangeCurrency: Currency.EUR,
      onAmountChange: () => {},
      onWalletChange: () => {},
    }
    const switcher = <WalletView type={type} {...defaultProps} {...props} />
    return shallow(switcher)
  }

  it('should render source WalletView', () => {
    const wrapper = createWalletView()
    expect(wrapper).to.matchSnapshot()
  })

  it('should render target WalletView', () => {
    const wrapper = createWalletView({}, 'target')
    expect(wrapper).to.matchSnapshot()
  })

  it('should handle wallet amount change', () => {
    const onAmountChange = spy()
    const wrapper = createWalletView({ onAmountChange })

    const props: any = wrapper.find('ConversionPart').props()
    props.onAmountChangeRequest({ persist: () => {}, target: { value: '42' } })

    expect(onAmountChange).to.have.been.called
  })

  it('should not handle wallet amount change for invalid input', () => {
    const onAmountChange = spy()
    const wrapper = createWalletView({ onAmountChange })

    const props: any = wrapper.find('ConversionPart').props()
    props.onAmountChangeRequest({
      persist: () => {},
      target: { value: '42asd' },
    })

    expect(onAmountChange).to.not.have.been.called
  })

  it('should handle wallet switch click', () => {
    const onWalletChange = spy()
    const wrapper = createWalletView({ onWalletChange })

    const props: any = wrapper.find('Switcher').props()
    props.onSwitchRequest(Currency.USD)

    expect(onWalletChange).to.have.been.calledOnceWith(Currency.USD)
  })
})
