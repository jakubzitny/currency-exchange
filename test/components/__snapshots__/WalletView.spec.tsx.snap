// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletView should render source WalletView 1`] = `ShallowWrapper {}`;

exports[`WalletView should render target WalletView 1`] = `ShallowWrapper {}`;
